app/controllers/users_controller.rb

```diff
@@ -209,6 +209,62 @@ def check_emails
     render json: failed_json, status: 403
   end
 
+  def update_primary_email
+    params.require(:email)
+
+    user = fetch_user_from_params
+    guardian.ensure_can_edit!(user)
+
+    old_primary = user.primary_email
+    if old_primary.email == params[:email]
+      return render json: success_json
+    end
+
+    new_primary = user.user_emails.find_by(email: params[:email])
+    if new_primary.blank?
+      return render json: failed_json.merge(errors: [I18n.t("change_email.doesnt_exist")]), status: 428
+    end
+
+    User.transaction do
+      old_primary.update!(primary: false)
+      new_primary.update!(primary: true)
+    end
+
+    if current_user.staff? && current_user != user
+      StaffActionLogger.new(current_user).log_update_email(user, previous_value: old_primary.email, new_value: new_primary.email)
+    else
+      UserHistory.create(action: UserHistory.actions[:update_email], target_user_id: user.id, previous_value: old_primary.email, new_value: new_primary.email)
```

Comment:
```suggestion
      UserHistory.create!(action: UserHistory.actions[:update_email], target_user_id: user.id, previous_value: old_primary.email, new_value: new_primary.email)
```

Reply:
I feel like logging should be part of the same transaction as updating the primary status of the email address as well.